src/coreclr/vm/interoplibinterface_objc.cpp

```diff
@@ -11,20 +11,68 @@
 
 #include "interoplibinterface.h"
 
-BOOL QCALLTYPE ObjCWrappersNative::TrySetGlobalMessageSendCallbacks(
-    _In_ void* fptr_objc_msgSend,
-    _In_ void* fptr_objc_msgSend_fpret,
-    _In_ void* fptr_objc_msgSend_stret,
-    _In_ void* fptr_objc_msgSendSuper,
-    _In_ void* fptr_objc_msgSendSuper_stret)
+#include <pinvokeoverride.h>
+
+#define OBJC_MSGSEND "objc_msgSend"
+
+namespace
+{
+    BOOL s_msgSendOverridden = FALSE;
+    void* s_msgSendOverrides[ObjCWrappersNative::MsgSendFunction::Last + 1] = {0};
+
+    const char* ObjectiveCLibrary = "/usr/lib/libobjc.dylib";
+    const char* MsgSendEntryPoints[ObjCWrappersNative::MsgSendFunction::Last + 1] =
+    {
+        OBJC_MSGSEND,
+        OBJC_MSGSEND "_fpret",
+        OBJC_MSGSEND "_stret",
+        OBJC_MSGSEND "Super",
+        OBJC_MSGSEND "Super_stret",
+    };
+
+    const void* MessageSendPInvokeOverride(const char* libraryName, const char* entrypointName)
+    {
+        // All overrides are in libobjc
+        if (strcmp(libraryName, ObjectiveCLibrary) != 0)
+            return nullptr;
+
+        // All overrides start with objc_msgSend
+        if (strncmp(entrypointName, OBJC_MSGSEND, _countof(OBJC_MSGSEND) -1) != 0)
+            return nullptr;
+        
+        for (int i = 0; i < _countof(MsgSendEntryPoints); ++i)
+        {
+            void* funcMaybe = s_msgSendOverrides[i];
+            if (funcMaybe != nullptr
+                && strcmp(entrypointName, MsgSendEntryPoints[i]) == 0)
+            {
+                return funcMaybe;
+            }
+        }
+
+        return nullptr;
+    }
+}
+
+BOOL QCALLTYPE ObjCWrappersNative::TrySetGlobalMessageSendCallback(
+    _In_ MsgSendFunction msgSendFunction,
+    _In_ void* fptr)
 {
     QCALL_CONTRACT;
 
+    bool success;
+
     BEGIN_QCALL;
 
+    success = FastInterlockCompareExchangePointer(&s_msgSendOverrides[msgSendFunction], fptr, NULL) == NULL;
+
+    // Set override if we haven't already
+    if (success && FALSE == FastInterlockCompareExchange((LONG*)&s_msgSendOverridden, TRUE, FALSE))
```

Comment:
This is setting the override callback right? Let's comment on that and validate the enum passed in is within range. That could be asserted here instead and the actual error case handled in managed if desired.